### PR TITLE
feat: Use react window for Typeahead dropdown 

### PR DESCRIPTION
**Pull request by ChitlangeSahas**
Closes https://github.com/influxdata/clockface/issues/790

### Changes

Add react window to render the list in typeahead dropdown! 

Uploading Screen Recording 2022-07-02 at 12.10.40 PM.mov…

### Checklist
Check all that apply

- [ ] Updated documentation to reflect changes
- [ ] Added entry to top of Changelog with link to PR (not issue)
- [ ] Tests pass
- [ ] Peer reviewed and approved
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

File: src/Components/Dropdowns/Composed/MenuDropdown.tsx [hunk 1]
```diff
@@ -25,6 +25,7 @@ import {
   StandardFunctionProps,
 } from '../../../Types'
 import classnames from 'classnames'
+import '../ScrollBarStyles.scss'
```
Comment (ChitlangeSahas) [inline src/Components/Dropdowns/Composed/MenuDropdown.tsx:28]: This adds styles to the List component so that it matches the Clockface scrollbar styles.

Comment (ChitlangeSahas) [reply src/Components/Dropdowns/Composed/MenuDropdown.tsx:28]: @wdoconnell this will fix this weird scrollbar you saw. 👍 

image

Comment (ChitlangeSahas) [reply src/Components/Dropdowns/Composed/MenuDropdown.tsx:28]: after the styles applied:

Screen Shot 2022-07-02 at 12 40 46 PM

Comment (wdoconnell) [reply src/Components/Dropdowns/Composed/MenuDropdown.tsx:28]: @ChitlangeSahas Thanks for fixing the styles!

**Review by wdoconnell — APPROVED** [review]
Looks good to me. Thanks again for fixing the style bug.

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx [hunk 2]
```diff
@@ -70,16 +48,10 @@ export const TypeAheadDropDown: FC<OwnProps> = ({
...
-  testIdSuffix = 'typeAhead',
+  placeholderText = 'Select a Value',
   selectedOption = null,
   className,
-  largeListSearch = false,
-  largeListCeiling = 0,
   menuTheme = DropdownMenuTheme.Onyx,
-  buttonTestId = 'type-ahead-dropdown--button',
-  menuTestID = 'type-ahead-dropdown--menu',
-  itemTestIdPrefix = 'type-ahead-dropdown--item',
```
Comment (ChitlangeSahas) [inline src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:82]: less configurable options to make test-ids useable in the UI. Spoke with @hoorayimhelping about the reasoning behind this change.

TLDR: we didn't have a specific need for this ultra-configurable test id setup, if we need it in the future we can consider adding it. At the moment this was just making it hard to use it.

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx [hunk 3]
```diff
@@ -94,28 +66,41 @@ export const TypeAheadDropDown: FC<OwnProps> = ({
 
   const [selectIndex, setSelectIndex] = useState(-1)
   const [queryResults, setQueryResults] = useState(items)
-  const [menuOpen, setMenuOpen] = useState<MenuStatus>(MenuStatus.Closed)
-
+  const [menuStatus, setMenuStatus] = useState<MenuStatus>(MenuStatus.Closed)
```
Comment (ChitlangeSahas) [inline src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:69]: `menuStatus` seemed like a better name, `menuOpen` sounds like a boolean when its not (?)

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx [hunk 4]
```diff
@@ -94,28 +66,41 @@ export const TypeAheadDropDown: FC<OwnProps> = ({
...
+  const [userHasTyped, setUserHasTyped] = useState(false)
```
Comment (ChitlangeSahas) [inline src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:70]: this keeps track of whether the user has typed or not, this helps us display a full list of items when viewing the dropdown. Instead of this: 


We see the full list : 
Screen Shot 2022-08-01 at 1 47 40 PM

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx [hunk 5]
```diff
@@ -126,82 +112,56 @@ export const TypeAheadDropDown: FC<OwnProps> = ({
...
-  }, [items, typedValue])
-
-  /**
-   *  using a ref to hold an instance variable:  what was last typed,
-   *  because without this 'click to select' doesn't work.
-   *  (you click to select, which clicks out of the dropdown, so then the dropdown sets
-   * the text to what was last selected.  this works fine for a class component,
-   * but here it uses the stale state of what was previously selected.
-   * this way, what was selected is saved in the ref.)
-   */
```
Comment (ChitlangeSahas) [inline src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:138]: I didn't find this to be true, the dropdown works without needing to store a ref? Am I missing something? 

https://user-images.githubusercontent.com/18511823/182245883-111cc654-6ee4-4db8-84ae-bf8588ca0e7a.mov

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx [hunk 6]
```diff
@@ -19,7 +12,8 @@ import {
 
 import {Input} from '../../Inputs/Input'
 import {DropdownHeader} from '../DropdownHeader'
-
+import {FixedSizeList as List} from 'react-window'
```
Comment (wdoconnell) [inline src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:15]: Is there a reason for using an alias here (List instead of FixedSizeList)?

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx (hunk 3 above)
Comment (wdoconnell) [reply src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:69]: LGTM.

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx (hunk 5 above)
Comment (wdoconnell) [reply src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:138]: I don't think you're missing anything: I see that this is working.  But do you know what the source of the rendering delay is here? It looks like there's a significant gap between the user selecting an option, and it registering in the typeahead.

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx [hunk 7]
```diff
@@ -1,12 +1,5 @@
...
-  ChangeEvent,
-  FC,
-  useEffect,
-  useMemo,
-  useRef,
-  useState,
-} from 'react'
+import React, {ChangeEvent, FC, useEffect, useMemo, useState} from 'react'
 import classnames from 'classnames'
 import {Dropdown} from '../.'
```
Comment (wdoconnell) [inline src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:4]: We could use absolute filepaths for the imports here.

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx [hunk 8]
```diff
@@ -34,10 +28,6 @@ interface OwnProps extends StandardFunctionProps {
   testIdSuffix?: string
   /**optional pre-selected option, must match exactly (name and id) an item in the items array */
   selectedOption?: SelectableItem | null
-  /**enables forced searching once dropdown list exceeds largeListSearch value */
-  largeListSearch?: boolean
-  /**the number of total items in the dropdown list before search is forced */
-  largeListCeiling?: number
```
[the remark this answers is not in the document]
Comment (wdoconnell) [reply src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:40]: 👍

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx (hunk 4 above)
Comment (wdoconnell) [reply src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:70]: I didn't understand this one: in the second image, why are options other than the options containing the text 'Los Angel' shown?

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx [hunk 9]
```diff
@@ -94,28 +66,42 @@ export const TypeAheadDropDown: FC<OwnProps> = ({
...
-  const [typedValue, setTypedValue] = useState<string>(initialTypedValue)
+  const [inputValue, setInputValue] = useState<string>(initialInputValue)
+  const [backupSelectionValue, setBackupSelectionValue] = useState<string>(
+    initialInputValue
+  )
 
-  const largeListValidation =
-    largeListSearch && queryResults.length > largeListCeiling
+  useEffect(() => {
+    // If selectedOption is invalid (not in the list of items), then we want to clear the input value
```
Comment (wdoconnell) [inline src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:87]: Does this happen? How would the user select an option that is not in the current list?

**Review by wdoconnell — CHANGES_REQUESTED** [review]
Thanks for doing this! I know this was a major undertaking.

Other than my comments, there could be some benefits to decomposition here - there are a lot of parts to this typeahead, so maybe it needs to be broken down into smaller functions.

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx (hunk 4 above)
Comment (ChitlangeSahas) [reply src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:70]: There are two scenarios to consider here:

When the user hasn't typed anything (input has whatever previous value it had), when they click the dropdown in that state, we want to show the entire list and not filter it by the typed value. 

When the user starts typing, we want to then start filtering.

Comment (ChitlangeSahas) [reply src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:70]: The second image is the user just clicking on the dropdown, and has not begun typing anything in it yet.

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx (hunk 5 above)
Comment (ChitlangeSahas) [reply src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:138]: That delay is just the dashboard changing the variable in all the places, not with the typeahead dropdown 👍

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx (hunk 6 above)
Comment (ChitlangeSahas) [reply src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:15]: I can't seem to remember my justification 😅 I'll change it back to FixedSizeList. 👍

File: src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx (hunk 9 above)
Comment (ChitlangeSahas) [reply src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx:87]: selectedOption is a prop so they could technically send in an invalid default selection. I noticed this happened in the Ui. So I thought I could handle it here.

**Review by wdoconnell — APPROVED** [review]
Looks good to me. Great work!
